File: affiliate_api/app/infrastructure/geolocation/ipinfo_service.py
# Source: AffiliateMarketing.Infrastructure/Services/IpInfoService.cs
import httpx
from dataclasses import dataclass
from app.core.config import settings

@dataclass
class GeoLocationData:
    country: str
    country_code: str
    city: str
    region: str
    timezone: str

LOCAL_IPS = {"::1", "127.0.0.1"}

class IpInfoService:
    def __init__(self):
        self._token = settings.IPINFO_TOKEN

    def _is_local(self, ip: str) -> bool:
        return ip in LOCAL_IPS or ip.startswith("192.168.") or ip.startswith("10.")

    async def get_location(self, ip_address: str) -> GeoLocationData:
        clean_ip = ip_address.split(":")[0].strip()
        if self._is_local(clean_ip):
            return GeoLocationData("Localhost", "LOC", "Development", "DevRegion", "UTC")
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                resp = await client.get(f"https://ipinfo.io/{clean_ip}?token={self._token}")
                resp.raise_for_status()
                data = resp.json()
                return GeoLocationData(
                    country=data.get("country", "Unknown"),
                    country_code=data.get("country", "XX"),
                    city=data.get("city", "Unknown"),
                    region=data.get("region", "Unknown"),
                    timezone=data.get("timezone", "UTC"),
                )
        except Exception:
            return GeoLocationData("Fallback", "XX", "Unknown", "Unknown", "UTC")

File: AffiliateMarketing.Infrastructure/Services/IpInfoService.cs
using System.Net.Http.Json;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using AffiliateMarketing.Application.Common.Interfaces;

namespace AffiliateMarketing.Infrastructure.Services.Geolocation;

public class IpInfoService : IGeoLocationService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<IpInfoService> _logger;
    private readonly string _token;

    public IpInfoService(
        HttpClient httpClient,
        ILogger<IpInfoService> logger,
        IConfiguration config)
    {
        _httpClient = httpClient;
        _logger = logger;
        // Superior Quality: Token is fetched from appsettings.json
        _token = config["IpInfo:Token"] ?? throw new ArgumentNullException("IpInfo Token is missing in configuration!");
    }

    public async Task<GeoLocationData> GetLocationAsync(string ipAddress, CancellationToken ct = default)
    {
        try
        {
            // 1. Sanitize the IP (Remove port if present)
            var cleanIp = ipAddress.Split(':')[0].Trim();

            // 2. Antigravity Check: Handle local development environments
            if (IsLocal(cleanIp))
            {
                return new GeoLocationData("Localhost", "LOC", "Development", "DevRegion", "UTC");
            }

            // 3. Make the API Call
            var url = $"https://ipinfo.io/{cleanIp}?token={_token}";
            var response = await _httpClient.GetFromJsonAsync<IpInfoResponse>(url, ct);

            // 4. Return clean data
            return new GeoLocationData(
                Country: response?.Country ?? "Unknown",
                CountryCode: response?.Country ?? "XX",
                City: response?.City ?? "Unknown",
                Region: response?.Region ?? "Unknown",
                Timezone: response?.Timezone ?? "UTC"
            );
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Antigravity Alert: Geolocation failed for IP {Ip}", ipAddress);
            return new GeoLocationData("Fallback", "XX", "Unknown", "Unknown", "UTC");
        }
    }

    private bool IsLocal(string ip) =>
        ip == "::1" || ip == "127.0.0.1" || ip.StartsWith("192.168.") || ip.StartsWith("10.");

    // Internal class to map the IpInfo JSON response
    private class IpInfoResponse
    {
        public string? Country { get; set; }
        public string? City { get; set; }
        public string? Region { get; set; }
        public string? Timezone { get; set; }
    }
}